scylla/src/transport/topology.rs

```diff
@@ -51,79 +57,95 @@ pub enum Strategy {
+        // - send REGISTER message to receive server events
+        // - send received events via server_event_sender
+        connection_config.event_sender = Some(server_event_sender);
+
+        let control_connection = ConnectionKeeper::new(
+            control_connection_address,
+            connection_config.clone(),
+            None,
+            None,
+            None,
+        );
 
         TopologyReader {
-            control_connections,
-            connect_port,
+            control_connection_address,
+            control_connection,
             connection_config,
+            known_peers: known_peers.into(),
         }
     }
 
     /// Fetches current topology info from the cluster
     pub async fn read_topology_info(&mut self) -> Result<TopologyInfo, QueryError> {
-        let result = self.fetch_topology_info().await;
+        let mut result = self.fetch_topology_info().await;
 
-        if let Ok(topology_info) = &result {
-            self.update_control_connections(topology_info);
-        }
+        // if fetching topology info on current control connection failed,
+        // try to fetch topology info from other known peer
+        for peer in self.known_peers.iter() {
+            if result.is_ok() {
+                break;
+            }
 
-        return result;
-    }
+            warn!(
+                control_connection_address = self.control_connection_address.to_string().as_str(),
+                error = result.as_ref().err().unwrap().to_string().as_str(),
+                "Falied to fetch topology info using current control connection"
+            );
 
-    async fn fetch_topology_info(&self) -> Result<TopologyInfo, QueryError> {
-        // TODO: Timeouts? New attempts in parallel?
+            self.control_connection_address = *peer;
+            self.control_connection = ConnectionKeeper::new(
+                self.control_connection_address,
+                self.connection_config.clone(),
+                None,
+                None,
+                None,
+            );
 
-        let mut last_error: QueryError = QueryError::ProtocolError(
-            "Invariant broken - TopologyReader control_connections not empty",
-        );
+            result = self.fetch_topology_info().await;
+        }
 
-        for conn in self.control_connections.values() {
-            match query_topology_info(conn, self.connect_port).await {
-                Ok(info) => return Ok(info),
-                Err(e) => last_error = e,
-            };
+        match &result {
+            Ok(topology_info) => {
+                self.update_known_peers(topology_info);
+                debug!("Fetched new topology info");
+            }
+            Err(error) => error!(
+                error = error.to_string().as_str(),
+                "Could not fetch topology info"
+            ),
         }
 
-        Err(last_error)
+        return result;
```

Comment:
Nit: no need to `return result;`, just `result` is sufficient and more idiomatic.

Reply:
Done